llpc/translator/lib/SPIRV/SPIRVReader.cpp

```diff
@@ -2215,6 +2215,33 @@ template <> Value *SPIRVToLLVM::transValueWithOpcode<OpLoad>(SPIRVValue *const s
 
   SPIRVType *const spvLoadType = spvLoad->getSrc()->getType();
 
+  // We like to check if accesses to local arrays with dynamic indices are out of bounds.
+  bool shouldGenerateBoundaryCheck = false;
+  SPIRVToLLVM::ScratchBoundaryCheckInfo boundaryCheckInfo;
+  std::tie(shouldGenerateBoundaryCheck, boundaryCheckInfo) = prepareScratchBoundaryCheck(spvLoad);
+
+  if (shouldGenerateBoundaryCheck) {
+    getBuilder()->CreateCondBr(boundaryCheckInfo.finalCmpResult, boundaryCheckInfo.intermediateBlock,
+                               boundaryCheckInfo.finalBlock);
+
+    // Add the load instruction to the load block,
+    // followed by an unconditional jump.
+    getBuilder()->SetInsertPoint(boundaryCheckInfo.intermediateBlock);
+    LoadInst *loadInst = dyn_cast<LoadInst>(addLoadInstRecursively(spvLoadType->getPointerElementType(), loadPointer,
+                                                                   isVolatile, isCoherent, isNonTemporal));
+    getBuilder()->CreateBr(boundaryCheckInfo.finalBlock);
+
+    // Let the load and the NOP converge and return the value.
+    getBuilder()->SetInsertPoint(boundaryCheckInfo.finalBlock);
+    PHINode *convergePhi = getBuilder()->CreatePHI(loadInst->getType(), 2);
+    convergePhi->addIncoming(Constant::getNullValue(loadInst->getType()), boundaryCheckInfo.entryBlock);
+    convergePhi->addIncoming(loadInst, boundaryCheckInfo.intermediateBlock);
+
+    return convergePhi;
+  }
+
+  // If no OOB guard needs to be inserted,
+  // just return the actual load instruction.
```

Comment:
nit: put these comments on the same line and let `clang-format` break it if it gets too long